### PR TITLE
Updates to nodejs/webapi project

### PR DESCRIPTION
**Pull request by juliamuiruri4**
## Purpose

* Solves 'You do not have permission to view this directory' error for the deployed express api

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone https://github.com/juliamuiruri4/APICenter-Reference
```
```
cd APICenter-Reference
```
```
azd up
```
Access nodejs api via [app-###-node-reference]

* Test the exercise

```
cd APICenter-Reference
```
Follow the provided steps - https://github.com/juliamuiruri4/APICenter-Reference/blob/main/docs/api-center-analyzer-integration-nodejs.md

## What to Check
Verify that the following are valid:-
* nodejs/webapi successfully deploys to an app service - app-###-node-reference
image

* json weather data is rendered at app-###-node-reference/weatherforecast
image

* swagger docs is rendered at app-###-node-reference/api-docs/swagger
image

### CODE REVIEW

File: infra/core/host/appservice.bicep [hunk 1]
```diff
@@ -32,7 +32,8 @@ param functionAppScaleLimit int = -1
 param linuxFxVersion string = runtimeNameAndVersion
 param minimumElasticInstanceCount int = -1
 param numberOfWorkers int = -1
-param scmDoBuildDuringDeployment bool = false
+param scmDoBuildDuringDeployment bool = true
```
Comment (justinyoo) [inline infra/core/host/appservice.bicep:35]: We should avoid this value being `true` due to the security reason

**Review by justinyoo — CHANGES_REQUESTED** [review]

**Review by justinyoo — CHANGES_REQUESTED** [review]
@juliamuiruri4 We have to update our npm packages due the security issue:

image

File: nodejs/.gitkeep [hunk 2]
```diff
@@ -0,0 +1 @@
+bin
```
Comment (justinyoo) [inline nodejs/.gitkeep:1]: The `.gitkeep` file doesn't do anything but placeholder to keep the directory structure. If you want to include that `bin` directory, please use `.gitignore`.

You can safely remove this `.gitkeep` file.

File: nodejs/webapi/public/stylesheets/style.css
Comment (justinyoo) [inline nodejs/webapi/public/stylesheets/style.css:1]: Do we need this file? It doesn't seem to be referenced from anywhere.

**Review by justinyoo — CHANGES_REQUESTED** [review]

**Comment by juliamuiruri4** [comment]
> @juliamuiruri4 We have to update our npm packages due the security issue:
> 
> image

Fixed

File: nodejs/webapi/routes/openapiservice.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+// Assuming `process.env.PORT` contains your port or named pipe value
+const portOrPipe = process.env.PORT || 3030; // Default to 3030 if not set
+const baseUrl = formatBaseUrl(portOrPipe);
+
+const environment = process.env.NODE_ENV || "development";
+
+let description, apis;
+
+if (environment === "development") {
+  description = `<a href="${baseUrl}/api-docs/swagger.json" target="_blank">${baseUrl}/api-docs/swagger.json</a>`;
```
Comment (justinyoo) [inline nodejs/webapi/routes/openapiservice.js:31]: Can it be markdown format?

File: nodejs/webapi/routes/openapiservice.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  apis = ["./app.js"];
+} else {
+  description = "An API that predicts upcoming weather";
```
Comment (justinyoo) [inline nodejs/webapi/routes/openapiservice.js:34]: We still need the swagger doc navigation link here.

File: nodejs/webapi/routes/openapiservice.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  apis: apis,
+};
+
+// Improved swagger definition
+const improvedSwaggerDefinition = {
+  definition: {
+    openapi: "3.0.1",
+    info: {
+      title: "Weather Forecast API",
+      description: description + "<br><br>An API that predicts upcoming weather",
```
Comment (justinyoo) [inline nodejs/webapi/routes/openapiservice.js:57]: It seems that `An API that predicts upcoming weather` will be displaying twice

**Review by justinyoo — CHANGES_REQUESTED** [review]

**Comment by juliamuiruri4** [comment]
> @juliamuiruri4 The app is not starting with the command, `npm start` or `npm run start`
> 
> image

Fixed.
image

File: nodejs/webapi/routes/openapiservice.js (hunk 3 above)
Comment (justinyoo) [reply nodejs/webapi/routes/openapiservice.js:31]: Yeah, because of the `target`, that makes sense.

File: nodejs/webapi/public/stylesheets/style.css
Comment (justinyoo) [reply nodejs/webapi/public/stylesheets/style.css:1]: Because this app is backend API, do we still need this?

**Review by justinyoo — CHANGES_REQUESTED** [review]

File: nodejs/webapi/routes/openapiservice.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  apis: ["./app.js"],
+};
+
+// Improved swagger definition
+const improvedSwaggerDefinition = {
+  definition: {
+    openapi: "3.0.1",
+    info: {
+      title: "Weather Forecast API",
+      description: `<a href="${baseUrl}/api-docs/swagger.json" target="_blank">${baseUrl}/api-docs/swagger.json</a>`,
```
Comment (justinyoo) [inline nodejs/webapi/routes/openapiservice.js:44]: We should also include `An API that predicts upcoming weather` in the description.

File: nodejs/webapi/app.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+app.use(cookieParser());
+app.use(express.static(path.join(__dirname, "public")));
+
+app.use("/weatherforecast", weatherRouter);
+
+if (process.env.NODE_ENV !== 'production') {
+  // route to serve basic/ improved JSON
+  app.get("/api-docs/swagger.json", (req, res) => {
+    // res.send(swaggerSpecs.improved);
+    res.send(swaggerSpecs.basic);
```
Comment (justinyoo) [inline nodejs/webapi/app.js:60]: Can we swap the location - basic to the top and improved to the bottom?

File: nodejs/webapi/app.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  });
+  
+  // setupSwaggerUi(app, swaggerSpecs.improved, "/api-docs/swagger");
+  setupSwaggerUi(app, swaggerSpecs.basic, "/api-docs/swagger");
```
Comment (justinyoo) [inline nodejs/webapi/app.js:64]: Can we swap the location - basic to the top and improved to the bottom?

**Review by justinyoo — CHANGES_REQUESTED** [review]

File: nodejs/webapi/routes/openapiservice.js (hunk 6 above)
Comment (juliamuiruri4) [reply nodejs/webapi/routes/openapiservice.js:44]: Done

File: nodejs/webapi/app.js (hunk 7 above)
Comment (juliamuiruri4) [reply nodejs/webapi/app.js:60]: Done

File: nodejs/webapi/app.js (hunk 8 above)
Comment (juliamuiruri4) [reply nodejs/webapi/app.js:64]: Done

**Review by justinyoo — APPROVED** [review]
Thaks for the update!

**Pull request merged**